plugins/pixart-tp/fu-pxi-tp-device.c

```diff
@@ -0,0 +1,1321 @@
+/*
+ * Copyright 2025 Harris Tai <harris_tai@pixart.com>
+ * Copyright 2025 Micky Hsieh <micky_hsieh@pixart.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include "fu-pxi-tp-common.h"
+#include "fu-pxi-tp-device.h"
+#include "fu-pxi-tp-firmware.h"
+#include "fu-pxi-tp-fw-struct.h"
+#include "fu-pxi-tp-register.h"
+#include "fu-pxi-tp-tf-communication.h"
+
+struct _FuPxiTpDevice {
+	FuHidrawDevice parent_instance;
+	guint8 sram_select;
+	guint8 ver_bank;
+	guint16 ver_addr;
+};
+
+G_DEFINE_TYPE(FuPxiTpDevice, fu_pxi_tp_device, FU_TYPE_HIDRAW_DEVICE)
+
+/* ---- reset mode & keys ---- */
+
+typedef enum {
+	PXI_TP_RESET_MODE_APPLICATION = 0,
+	PXI_TP_RESET_MODE_BOOTLOADER = 1,
+} FuPxiTpResetMode;
+
+enum {
+	PXI_TP_RESET_KEY1_MAGIC = 0xaa,
+	PXI_TP_RESET_KEY2_REGULAR = 0xbb,
+	PXI_TP_RESET_KEY2_BOOTLOADER = 0xcc,
+};
+
+/* ---- flash command constants ---- */
+
+enum {
+	PXI_TP_FLASH_INST_CMD0 = 0x00,
+	PXI_TP_FLASH_INST_CMD1 = 0x01,
+};
+
+enum {
+	PXI_TP_FLASH_EXEC_STATE_BUSY = 0x01,
+	PXI_TP_FLASH_EXEC_STATE_SUCCESS = 0x00,
+	PXI_TP_FLASH_WRITE_ENABLE_SUCCESS = 0x02,
+	PXI_TP_FLASH_STATUS_BUSY = 0x01, /* busy bit in status register */
+};
+
+enum {
+	PXI_TP_FLASH_CCR_WRITE_ENABLE = 0x00000106,
+	PXI_TP_FLASH_CCR_READ_STATUS = 0x01000105,
+	PXI_TP_FLASH_CCR_ERASE_SECTOR = 0x00002520,
+	PXI_TP_FLASH_CCR_PROGRAM_PAGE = 0x01002502,
+};
+
+/* ---- flash properties ---- */
+enum {
+	PXI_TP_SECTOR_SIZE = 4096,
+	PXI_TP_PAGE_SIZE = 256,
+	PXI_TP_PAGES_COUNT_PER_SECTOR = 16,
+};
```

Comment:
i don't think this is an enum -- they're just three # define values. enums are supposed to form a "set" if you see what I mean.